xilem_core/src/element.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! The types which can be used as elements in a [`View`](crate::View)
+
+/// A type which can be used as the `Element` associated type for a [`View`](crate::View).
+///
+/// It is expected that most libraries using `xilem_core` will have a generic
+/// implementation of this trait for their widget type.
+/// Additionally, this may also be implemented for other types, depending on the
+/// needs of the specific parent view.
+/// In Xilem (the user interface library), this is also used for types containing the
+/// flex properties of their child views, and window properties.
+///
+// /// This does require the reference type to be reborrowable.
+///
+/// In most cases, there will be a corresponding implementation of [`SuperElement<Self>`] for
+/// some other type.
+/// This will be the generic form of this type, which is used for the implementation of [`AnyView`].
+///
+/// [`AnyView`]: crate::AnyView
+///
+// TODO: Rename so that it doesn't conflict with the type parameter names
+pub trait ViewElement {
+    /// The reference form of this `Element` for editing.
+    ///
+    /// This is provided to [`View::rebuild`](crate::View::rebuild) and
+    /// [`View::teardown`](crate::View::teardown).
+    /// This enables greater flexibility in the use of the traits, such as
+    /// for reference types which contain access to parent state.
+    type Mut<'a>;
+    /* /// Perform a reborrowing access to the reference type, which allows re-using the reference
+    /// even if it gets passed to another function.
+    ///
+    /// This is the more general form of [`with_reborrow`](ViewElement::with_reborrow).
+    /// See its documentation for more details.
+    fn with_reborrow_val<R: 'static>(
+        this: Self::Mut<'_>,
+        f: impl FnOnce(Self::Mut<'_>) -> R,
+    ) -> (Self::Mut<'_>, R);
+
+    /// Perform a reborrowing access to the reference type, which allows re-using the reference
+    /// even if it gets passed to another function.
+    ///
+    /// The closure accepts a second parameter of type `&()` because of rustc issue [#49601].
+    /// In this case, the diagnostic is probably incorrect.
+    /// When calling this function, you can safely ignore this parameter (i.e. `|element, _| {...}`).
+    ///
+    /// If you need to get a return value from the .
+    /// Unfortunately, it isn't possible to abstract over reborrowing without a closure or equivalent.
+    ///
+    /// [#49601](https://github.com/rust-lang/rust/issues/49601)
+    fn with_reborrow(this: Self::Mut<'_>, f: impl FnOnce(Self::Mut<'_>)) -> Self::Mut<'_> {
+        let (this, ()) = Self::with_reborrow_val(this, f);
+        this
+    } */
+}
```

Comment:
IMO reborrowing isn't unidiomatic.

This API form is needed because it's the only way to represent reborrowing in the type system.
But yes, Masonry makes it complicated